basic document repository implementation

slu-it opened the pull request:
Includes:
- configuration of repositories 
- definiton of repository and repository factory API
- management of repositories

There are two new modules:
- `livingdoc-repositories` defines API classes and provides manages repository instances and configuration
- `livingdoc-repository-local` base implementation for a local repository intended for HTML, Markdown, AsciiDoc etc. documents checked in with the code

The idea for configuring repositories is to define them in the central `livingdoc.yml` files with all the other configurations. Each repository entry specifies a _unique_ name, the corresponding factory class and a repository specific configuration section.

bitterblue commented on livingdoc-repositories/src/main/kotlin/org/livingdoc/repositories/config/ConfigUtils.kt:
```diff
@@ -0,0 +1,29 @@
+    fun loadFileFromClasspath(filePath: String): Map<String, Any> {
+        val file = getFile(filePath) ?: throw FileNotFoundException("File not found: $filePath")
+        return file.openStream().use {
+            yaml.load(it) as Map<String, Any>
```
Why the cast?
Also: zero warning policy :)

slu-it replied:
I'll suppress the warning, cast is necessary because of "not so good" API of Snakeyml

bitterblue commented on livingdoc-repositories/src/main/kotlin/org/livingdoc/repositories/RepositoryManager.kt:
```diff
@@ -0,0 +1,57 @@
+    class RepositoryAlreadyRegisteredException(name: String, repository: DocumentRepository)
+        : RuntimeException("Document repository with name '$name' already registered! Cant register instance of ${repository.javaClass.canonicalName}")
+
+    class NoRepositoryFactoryException(name: String)
```
I didn't find tests for the last two exceptions. Shouldn't we check for that?

slu-it replied:
Yes, there is at least one missing test.

bitterblue commented on livingdoc-repositories/src/main/kotlin/org/livingdoc/repositories/config/Configuration.kt:
```diff
@@ -0,0 +1,20 @@
+            return loadFromClassPathFile("livingdoc.yml")
+        }
+
+        fun loadFromClassPathFile(filePath: String): Configuration {
```
I'd turn this into `loadFromStream` and keep the knowledge about files and classpaths in `load()`. That gives us more options later and also would allow to specify yaml directly in the test file.

slu-it replied:
Ok, I'll do that.

slu-it commented on livingdoc-repositories/src/main/kotlin/org/livingdoc/repositories/config/ConfigUtils.kt:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    fun <T : Any> toObject(configData: Map<String, Any>, type: KClass<T>): T {
+        val yamlString = yaml.dump(configData)
```
Necessary for reading parts of the original YAML structure (e.g. the `config` node) as objects. There is no read `Map<String, Any>` as object method.

slu-it commented:
@bitterblue I've changed what we discussed regarding your review. I also refactored some more regarding the configuration as can be seen in `YamlUtils` and the `RepositoryManagerTest`.
Can you review again? I'll squash the commits as soon as I get the green light :)

The two new commits are prefixed with `fixup!`, it would be nice to know if a "rebase and merge" from the GitHub UI would actually fixup the commits while rebasing ...

slu-it commented:
https://help.github.com/articles/about-pull-request-merges/ does not mention fixup commits explicitly - I'll just rebase manually and update the PR as soon as the reviews are done.

bitterblue commented on livingdoc-repository-local/src/main/kotlin/org/livingdoc/repositories/local/LocalRepository.kt:
```diff
@@ -10,7 +9,7 @@ class LocalRepository(
 
     override fun getDocument(documentIdentifier: String): Document? {
-        throw DocumentNotFoundException(name, documentIdentifier)
+        throw IllegalStateException("not implemented")
```
IllegalStateException in Kotlin is simply a call to `error()`.
